[PT2 Compiler] Add annotation for dynamo disabled callables

malaybag opened the pull request:
Summary: To make torch.export compatible with PT2 compile (which is done on top of exported model) we need to store torch._dynamo.disable attributes in exported model and later restore this after unflattening of exported model. This diff will add annotations to all nodes with torch._dynamo.disable, which will be preserved during exporting.

Test Plan:
```
buck test mode/opt caffe2/test:test_export -- 'test_dynamo_disable_annotations'
```
https://www.internalfb.com/intern/testinfra/testrun/6473924770741560

Differential Revision: D85302730


cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @amjames @Lucaskabela

angelayi reviewed (commented):
@williamwen42 do you mind taking a look at the dynamo changes

williamwen42 reviewed (commented):
made comments on dynamo parts

malaybag commented on torch/_dynamo/eval_frame.py:
```diff
@@ -1133,7 +1133,18 @@ def _fn(*args: Any, **kwargs: Any) -> Any:
                 _maybe_set_eval_frame(_callback_from_stance(self.callback))
                 try:
-                    return fn(*args, **kwargs)
+                    with fx_traceback.annotate(
```
Sorry, n00b questions, do we need to trace the with block? The goal here is just to update custom meta, which the is done by fx_traceback.annotate.

angelayi commented on torch/_dynamo/external_utils.py:
```diff
@@ -196,7 +197,18 @@ def get_nonrecursive_disable_wrapper(fn: Callable[_P, _R]) -> Callable[_P, _R]:
     def nonrecursive_disable_wrapper(*args: _P.args, **kwargs: _P.kwargs) -> _R:
-        return fn(*args, **kwargs)
+        with fx_traceback.annotate(
+            {
```
@malaybag since we decided to not support recursive=False, let's just throw a RuntimeError here if torch.compiler.is_exporting()?

williamwen42 reviewed (commented):
ok lgtm on dynamo side

malaybag committed:
[PT2 Compiler] Add annotation for dynamo disabled callables (#166341)

Summary:

To make torch.export compatible with PT2 compile (which is done on top of exported model) we need to store torch._dynamo.disable attributes in exported model and later restore this after unflattening of exported model. This diff will add annotations to all nodes with torch._dynamo.disable, which will be preserved during exporting.

Test Plan:
```
buck test mode/opt caffe2/test/dynamo:test_dynamo -- 'test_dynamo_disable_annotations'
```
https://www.internalfb.com/intern/testinfra/testrun/562950408912605

Reviewed By: angelayi, williamwen42

Differential Revision: D85302730
